File: Source/Parameters.cpp
#include "Parameters.h"

namespace openfad::params
{
namespace
{
using APVTS = juce::AudioProcessorValueTreeState;

std::unique_ptr<juce::AudioParameterFloat> floatParameter (const char* parameterId,
                                                            const char* name,
                                                            float minimum,
                                                            float maximum,
                                                            float defaultValue,
                                                            const char* unit = {})
{
    juce::NormalisableRange<float> range (minimum, maximum);
    return std::make_unique<juce::AudioParameterFloat> (parameterId, name, range, defaultValue, unit);
}

std::unique_ptr<juce::AudioParameterChoice> choiceParameter (const char* parameterId,
                                                              const char* name,
                                                              const juce::StringArray& choices,
                                                              int defaultIndex)
{
    return std::make_unique<juce::AudioParameterChoice> (parameterId, name, choices, defaultIndex);
}

std::unique_ptr<juce::AudioParameterBool> boolParameter (const char* parameterId,
                                                          const char* name,
                                                          bool defaultValue)
{
    return std::make_unique<juce::AudioParameterBool> (parameterId, name, defaultValue);
}
}

juce::AudioProcessorValueTreeState::ParameterLayout createParameterLayout()
{
    APVTS::ParameterLayout layout;

    layout.add (floatParameter (id::inputTrim, "Input Trim", -24.0f, 24.0f, 0.0f, "dB"));
    layout.add (floatParameter (id::outputTrim, "Output Trim", -24.0f, 12.0f, 0.0f, "dB"));
    layout.add (floatParameter (id::mix, "Mix", 0.0f, 1.0f, 0.35f));
    layout.add (boolParameter (id::bypass, "Bypass", false));
    layout.add (choiceParameter (id::quality, "Quality", qualityNames(), 0));

    layout.add (choiceParameter (id::model, "Speaker Model", modelNames(), 1));
    layout.add (boolParameter (id::modelBypass, "Model Bypass", false));
    layout.add (floatParameter (id::drive, "Drive", 0.0f, 1.0f, 0.2f));
    layout.add (floatParameter (id::resonance, "Cabinet Resonance", 0.0f, 1.0f, 0.35f));
    layout.add (floatParameter (id::damping, "Cabinet Damping", 0.0f, 1.0f, 0.5f));
    layout.add (boolParameter (id::loudnessMatch, "Loudness Match", true));

    layout.add (choiceParameter (id::structure, "Rotor Structure", structureNames(), 0));
    layout.add (choiceParameter (id::feedMode, "Feed Mode", feedModeNames(), 0));
    layout.add (choiceParameter (id::renderMode, "Render Mode", renderModeNames(), 0));
    layout.add (choiceParameter (id::speedMode, "Speed Mode", speedModeNames(), 1));
    layout.add (floatParameter (id::freeRate, "Free Rate", 0.02f, 20.0f, 0.8f, "Hz"));
    layout.add (choiceParameter (id::syncDivision, "Sync Division", syncDivisionNames(), 5));
    layout.add (floatParameter (id::inertia, "Inertia", 0.05f, 12.0f, 2.2f, "s"));
    layout.add (choiceParameter (id::direction, "Direction", juce::StringArray { "CW", "CCW" }, 0));
    layout.add (floatParameter (id::depth, "Rotor Depth", 0.0f, 1.0f, 0.75f));
    layout.add (floatParameter (id::distance, "Listener Distance", 0.5f, 3.0f, 1.2f, "m"));
    layout.add (floatParameter (id::angle, "Listener Angle", -45.0f, 45.0f, 0.0f, "deg"));
    layout.add (floatParameter (id::earlyReflections, "Early Reflections", 0.0f, 1.0f, 0.25f));
    layout.add (floatParameter (id::roomDamping, "Room Damping", 0.0f, 1.0f, 0.55f));
    layout.add (floatParameter (id::modelAmount, "Model Amount", 0.0f, 1.0f, 1.0f));
    layout.add (floatParameter (id::rotatorAmount, "Rotator Amount", 0.0f, 1.0f, 1.0f));

    layout.add (boolParameter (id::dreamBypass, "Dream Bypass", false));
    layout.add (floatParameter (id::predelay, "Dream Predelay", 0.0f, 0.25f, 0.035f, "s"));
    layout.add (boolParameter (id::predelaySync, "Predelay Sync", false));
    layout.add (floatParameter (id::diffusion, "Diffusion", 0.0f, 1.0f, 0.45f));
    layout.add (floatParameter (id::tail, "Dream Tail", 0.2f, 12.0f, 3.5f, "s"));
    layout.add (floatParameter (id::microshift, "Microshift", 0.0f, 25.0f, 8.0f, "cents"));
    layout.add (floatParameter (id::dreamDamping, "Dream Damping", 0.0f, 1.0f, 0.35f));
    layout.add (floatParameter (id::feedback, "Dream Feedback", 0.0f, 0.96f, 0.58f));
    layout.add (boolParameter (id::freeze, "Dream Freeze", false));

    layout.add (floatParameter (id::character, "Character", 0.0f, 1.0f, 0.35f));
    layout.add (floatParameter (id::motion, "Motion", 0.0f, 1.0f, 0.35f));
    layout.add (floatParameter (id::space, "Space", 0.0f, 1.0f, 0.3f));
    layout.add (floatParameter (id::dream, "Dream", 0.0f, 1.0f, 0.25f));

    // Append new parameters so existing host automation/index ordering remains stable.
    layout.add (floatParameter (id::dopplerAmount, "Doppler Amount", 0.0f, 1.0f, 1.0f));

    return layout;
}

juce::StringArray modelNames()
{
    return { "Pocket Radio", "Console Coax", "Cinema Horn", "British Shelf",
             "American Tower", "Nearfield Monitor", "PA Stack", "Modern Reference" };
}

juce::StringArray structureNames()
{
    return { "Horn + Drum", "Eccentric Ports", "Prism Diffuser" };
}

juce::StringArray feedModeNames()
{
    return { "Mono Sum", "Linked Stereo", "Dual Rotor" };
}

juce::StringArray renderModeNames()
{
    return { "Binaural", "Speaker Stereo" };
}

juce::StringArray speedModeNames()
{
    return { "Stop", "Slow", "Fast", "Free", "Sync" };
}

juce::StringArray qualityNames()
{
    return { "Live", "Studio" };
}

juce::StringArray syncDivisionNames()
{
    return { "1/32", "1/16", "1/8", "1/4", "1/2", "1 bar", "2 bars", "4 bars", "8 bars" };
}

} // namespace openfad::params

File: WebUI/src/garden/parameters.ts
import {
  defaultPluginValues, parameterHelp, modelOptions, structureOptions,
  feedModeOptions, renderModeOptions, speedModeOptions, syncDivisionOptions,
  qualityOptions, type PluginValues,
} from "../parameterContract";

export type Id = Exclude<keyof PluginValues, "bpm">;
export type Control = { id: Id; min: number; max: number; step: number; unit?: string; log?: boolean };
const c = (id: Id, min = 0, max = 1, step = .01, unit = "%", log = false): Control =>
  ({ id, min, max, step, unit, log });
export const macros = ["character", "motion", "space", "dream"] as const;
export const names: Partial<Record<Id, string>> = { character: "音色", motion: "运动", space: "空间", dream: "梦境" };
export const title = (id: Id) => names[id] ?? parameterHelp[id]?.cn ?? id;
export const colors = ["#c57169", "#699b89", "#659cae", "#c6a251", "#8484a4", "#778681"];
export const groups = [
  { name: "音箱", en: "SPEAKER", macro: "character", controls: [c("modelAmount"), c("drive"), c("resonance"), c("damping")], options: ["model", "modelBypass", "loudnessMatch"] },
  { name: "旋转", en: "ROTATION", macro: "motion", controls: [c("freeRate", .02, 20, .01, "Hz", true), c("inertia", .05, 12, .01, "s"), c("rotatorAmount"), c("depth")], options: ["speedMode", "syncDivision", "direction", "structure"] },
  { name: "空间", en: "SPACE", macro: "space", controls: [c("distance", .5, 3, .01, "m"), c("angle", -45, 45, .1, "°"), c("earlyReflections"), c("roomDamping")], options: ["feedMode", "renderMode"] },
  { name: "梦境", en: "DREAM", macro: "dream", controls: [c("predelay", 0, .25, .001, "s"), c("diffusion"), c("tail", .2, 12, .01, "s", true), c("microshift", 0, 25, .01, "ct"), c("dreamDamping"), c("feedback", 0, .96)], options: ["dreamBypass", "predelaySync", "freeze"] },
  { name: "多普勒", en: "DOPPLER", macro: "", controls: [c("dopplerAmount")], options: [] },
  { name: "输出", en: "OUTPUT", macro: "", controls: [c("inputTrim", -24, 24, .1, "dB"), c("mix"), c("outputTrim", -24, 12, .1, "dB")], options: ["quality", "bypass"] },
] as const;
export const choices: Partial<Record<Id, readonly string[]>> = {
  model: modelOptions, structure: structureOptions, feedMode: feedModeOptions,
  renderMode: renderModeOptions, speedMode: speedModeOptions, syncDivision: syncDivisionOptions,
  quality: qualityOptions, direction: ["CW", "CCW"],
};
export const controls: Control[] = [...macros.map(id => c(id)), ...groups.flatMap(g => [...g.controls])];
export const definitions = new Map(controls.map(control => [control.id, control]));
export const defaults = { ...defaultPluginValues };
export const clamp = (x: number, a = 0, b = 1) => Math.min(b, Math.max(a, x));
export function ratio(id: Id, value: number) {
  const d = definitions.get(id)!;
  return d.log ? Math.log(value / d.min) / Math.log(d.max / d.min) : (value - d.min) / (d.max - d.min);
}
export function fromRatio(id: Id, r: number) {
  const d = definitions.get(id)!;
  const value = d.log ? d.min * (d.max / d.min) ** clamp(r) : d.min + clamp(r) * (d.max - d.min);
  return clamp(Number((Math.round(value / d.step) * d.step).toFixed(6)), d.min, d.max);
}
export function format(id: Id, value: number | boolean) {
  if (typeof value === "boolean") return value ? "ON" : "OFF";
  if (choices[id]) return choices[id]![value];
  const d = definitions.get(id)!;
  return d.unit === "%" ? `${Math.round(value * 100)}%` :
    `${value.toFixed(d.step < .005 ? 3 : d.step < .05 ? 2 : 1)} ${d.unit}`;
}
export function disabled(id: Id, values: PluginValues) {
  return (id === "freeRate" && values.speedMode !== 3) || (id === "syncDivision" && values.speedMode !== 4);
}
